### PR TITLE
Make more base nodes thread safe

### PR DESCRIPTION
**Pull request by reduz**
Ongoing work to make more of the base nodes thread safe.

### CODE REVIEW

File: scene/3d/node_3d.cpp [hunk 1]
```diff
@@ -298,31 +313,41 @@ void Node3D::set_global_transform(const Transform3D &p_transform) {
...
-	if (data.dirty & DIRTY_GLOBAL_TRANSFORM) {
-		if (data.dirty & DIRTY_LOCAL_TRANSFORM) {
-			_update_local_transform();
+	/* Due to how threads work at scene level, while this global transform won't be able to be changed from outside a thread,
+	 * it is possible that multiple threads can access it while it's dirty from previous work. Due to this, we must ensure that
+	 * the dirty/update process is thread safe by utilizing atomic copies.
+	 */
+
+	if (data.dirty.get() & DIRTY_GLOBAL_TRANSFORM) {
+		if (data.dirty.get() & DIRTY_LOCAL_TRANSFORM) {
```
Comment (RandomShaper) [inline scene/3d/node_3d.cpp:334]:
```suggestion
	int dirty_local = data.dirty.get();
	if (dirty_local & DIRTY_GLOBAL_TRANSFORM) {
		if (dirty_local & DIRTY_LOCAL_TRANSFORM) {
```

**Comment by RandomShaper** [comment]
Great work.

There's something I'm not very fond of: non-group processing has to pay the cost of acq-rel. And it's hard to tell how bad that it versus the alternatives without benchmarking on various platforms. However, I'd still suggest that `dirty`, `xform_dirty`, etc. were treated as non-atomics, relaxedly during regular processing. Such dual identities could be achieved via type-punning; e.g.:
```
union {
    SafeFlag atomic; 
    bool regular;
} xform_dirty;
```

For `SafeNumeric` we already have `SAFE_NUMERIC_TYPE_PUN_GUARANTEES()` that could be leveraged. We could have a similar one for `bool`.

Not a blocker for this PR, but I strongly suggest to have it considered.

File: scene/2d/node_2d.h [hunk 2]
```diff
@@ -36,15 +36,14 @@
 class Node2D : public CanvasItem {
 	GDCLASS(Node2D, CanvasItem);
 
+	SafeFlag _xform_dirty;
```
Comment (RandomShaper) [inline scene/2d/node_2d.h:39]:
```suggestion
	SafeFlag xform_dirty;
```

File: scene/2d/node_2d.cpp [hunk 3]
```diff
@@ -142,35 +142,40 @@ void Node2D::reparent(Node *p_parent, bool p_keep_global_transform) {
...
 void Node2D::set_rotation_degrees(real_t p_degrees) {
+	ERR_THREAD_GUARD;
 	set_rotation(Math::deg_to_rad(p_degrees));
 }
 
 void Node2D::set_skew(real_t p_radians) {
-	if (_xform_dirty) {
+	ERR_THREAD_GUARD;
+	if (_xform_dirty.is_set()) {
 		const_cast<Node2D *>(this)->_update_xform_values();
```
[the remark this answers is not in the document]
Comment (reduz) [reply scene/2d/node_2d.cpp:170]: Can be done in a separate PR I guess.

**Comment by reduz** [comment]
@RandomShaper It makes sense, it would be good if you want to do this as a subsequent PR (for Control, Node2D and Node3D).

**Review by RandomShaper — APPROVED** [review]
I think this looks good enough to be put under test in the wild.

File: scene/main/canvas_item.cpp [hunk 4]
```diff
@@ -565,25 +601,30 @@ void CanvasItem::draw_dashed_line(const Point2 &p_from, const Point2 &p_to, cons
...
 void CanvasItem::draw_line(const Point2 &p_from, const Point2 &p_to, const Color &p_color, real_t p_width, bool p_antialiased) {
+	ERR_THREAD_GUARD;
 	ERR_FAIL_COND_MSG(!drawing, "Drawing is only allowed inside NOTIFICATION_DRAW, _draw() function or 'draw' signal.");
 
 	RenderingServer::get_singleton()->canvas_item_add_line(canvas_item, p_from, p_to, p_color, p_width, p_antialiased);
 }
 
 void CanvasItem::draw_polyline(const Vector<Point2> &p_points, const Color &p_color, real_t p_width, bool p_antialiased) {
+	ERR_THREAD_GUARD;
+	ERR_THREAD_GUARD;
```
Comment (bruvzg) [inline scene/main/canvas_item.cpp:612]:
```suggestion
	ERR_THREAD_GUARD;

```

File: scene/2d/node_2d.cpp [hunk 5]
```diff
@@ -29,7 +29,6 @@
 /**************************************************************************/
 
 #include "node_2d.h"
-
```
Comment (akien-mga) [inline scene/2d/node_2d.cpp:32]: No no.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**